---
title: "Quality Control"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Reports/Filtering.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

We loop trough the folders containing the filtered matrix files, barcodes and genenames to read in the data in form of SingleCellExperiment (sce) objects.

```{r libraries, message=FALSE}
library(scran)
library(scater)
library(DropletUtils)
library(ggplot2)
library(Matrix)
library(cowplot)
library(openxlsx)
```

```{r additional_files}
MT_genes <- read.table("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/MT_genes.txt", sep = "\t", header = TRUE)
```

```{r data, eval=FALSE}
folders <- list.files("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Data/", full.names = TRUE)
folder.names <- list.files("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Data/", full.names = FALSE)

# List to store sce objects
sce.list <- list()

# Read in the data
for(i in 1:length(folders)){
  cur_sce <- read10xCounts(folders[i])
  colData(cur_sce)$Patient <- as.character(sapply(folder.names, 
                                         function(n){unlist(strsplit(n, "_"))[1]})[i])
  colData(cur_sce)$Tissue <- as.character(sapply(folder.names, 
                                        function(n){unlist(strsplit(n, "_"))[3]})[i])
  sce.list[[folder.names[i]]] <- cur_sce
}

# Save list
saveRDS(sce.list, "../../../Dropbox (Cambridge University)/Oesophagus_single_cell/All_sce.rds")
```

Next we filter the cells based on several criteria. For this, we will plot these QC features first.

```{r QC_metrics}
sce.list <- readRDS("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/All_sce.rds")
for(i in sce.list){
  i <- suppressMessages(calculateQCMetrics(i))
  
  # Library size
  print(ggplot(as.data.frame(colData(i))) + 
    geom_point(aes(1:ncol(i), log10(total_counts))) +
    xlab("Cells"))
  
  # Number of genes detected
  print(ggplot(as.data.frame(colData(i))) + 
    geom_point(aes(1:ncol(i), total_features)) +
    xlab("Cells"))
  
  # Mitochondrial genes
  print(ggplot(data.frame(MT_genes = Matrix::colSums(counts(i)[rownames(i) %in%
                                                           MT_genes$Gene.stable.ID,])/
        Matrix::colSums(counts(i)))*100) +
    geom_point(aes(1:ncol(i), MT_genes)) + 
     ylab("% mitochondrial reads") + xlab("Cells"))
  
  # Marker gene expression
  # VIM
  print(ggplot(data.frame(VIM = log10(counts(i)["ENSG00000026025",] + 1))) +
    geom_point(aes(1:ncol(i), VIM)) + xlab("Cells") +
    ylab("log10[VIM]"))

  # PTPRC
  print(ggplot(data.frame(PTPRC = log10(counts(i)["ENSG00000081237",] + 1))) +
    geom_point(aes(1:ncol(i), PTPRC)) + xlab("Cells") +
    ylab("log10[VIM]"))
}

# Create xlsx file with entries for QC thresholds
df <- data.frame(names = folder.names,
           lower_total_counts = rep(0, length(folder.names)),
           upper_total_counts = rep(0, length(folder.names)),
           lower_total_features = rep(0, length(folder.names)),
           upper_total_features = rep(0, length(folder.names)),
           lower_mito = rep(0, length(folder.names)),
           upper_mito = rep(0, length(folder.names)),
           VIM = rep(0, length(folder.names)),
           PTPRC = rep(0, length(folder.names)))

write.xlsx(df, "../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Additional_files/QC_metrics.xlsx")
```