---
title: "Cell mapping using the Data from Owen et al."
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Reports/OwenMapping.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

The data can be obtained from the original puplication:
Owen, R. P. et al., Single cell RNA-seq reveals profound transcriptional similarity between Barrettâ€™s oesophagus and oesophageal submucosal glands, Nature Communications, 9, 2018

```{r}
library(scran)
library(SingleCellExperiment)
library(Matrix)
library(scater)
library(ggplot2)
source("../../Analysis/Functions/auxiliary.R")

# Owen read counts
owen.data <- read.table(url("https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-018-06796-9/MediaObjects/41467_2018_6796_MOESM8_ESM.txt"))

# Rename columns
colnames(owen.data) <- as.character(sapply(colnames(owen.data), function(n){paste(unlist(strsplit(n, "_"))[3:5], collapse = "_")}))

# Owen metadata
owen.meta <- read.table(url("https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-018-06796-9/MediaObjects/41467_2018_6796_MOESM9_ESM.txt"), header = TRUE)

# Reorder cells based on metadata
owen.data <- owen.data[,as.character(owen.meta$sample_id)]

# Our data
sce <- readRDS("../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/High_quality_sce.rds")
```

# Process Owen data

```{r processing}
# Remove genes that contain NA
owen.data <- owen.data[!is.na(rowSums(owen.data)),]

# Remove ERCC counts 
owen.data <- owen.data[!grepl("ERCC", rownames(owen.data)),]

# Build the SCE object
sce.owen <- SingleCellExperiment(assays = list(counts = as(as.matrix(owen.data), "dgCMatrix")),
                                 colData = as.data.frame(owen.meta))
# Remove low quality cells
sce.owen <- sce.owen[,apply(colData(sce.owen), 1, function(n){sum(is.na(n))/7}) != 1]

# Visualize quality
sce.owen <- calculateQCMetrics(sce.owen)

ggplot(as.data.frame(colData(sce.owen))) +
  geom_point(aes(total_features_by_counts, log10_total_counts, colour = tissue))
ggplot(as.data.frame(colData(sce.owen))) +
  geom_point(aes(total_features_by_counts, log10_total_counts, colour = all_k8))

# Remove cells with few genes expressed (blanks as these will distort the normalization)
sce.owen <- sce.owen[,colData(sce.owen)$total_features_by_counts > 1000]

# Normalize the data
cluster <- quickCluster(sce.owen, method = "igraph", irlba.args = c("work" = 100), 
                         max.size = 2000, min.size = 50)
sce.owen <- computeSumFactors(sce.owen, clusters=cluster)
sce.owen <- normalize(sce.owen, return_log = TRUE)
```

# Visualize owen data

```{r visualization}
# Compute tSNE
set.seed(1234)
sce.owen <- runTSNE(sce.owen)

# Visualize tissue type
ggplot(data.frame(tSNE1 = reducedDims(sce.owen)$TSNE[,1],
                  tSNE2 = reducedDims(sce.owen)$TSNE[,2],
                  colData(sce.owen))) + geom_point(aes(tSNE1, tSNE2, colour = tissue))

# Visualize clustering on all cells
ggplot(data.frame(tSNE1 = reducedDims(sce.owen)$TSNE[,1],
                  tSNE2 = reducedDims(sce.owen)$TSNE[,2],
                  colData(sce.owen))) + geom_point(aes(tSNE1, tSNE2, colour = as.factor(all_k8)))

# Visualize clustering of gland cells
ggplot(data.frame(tSNE1 = reducedDims(sce.owen)$TSNE[,1],
                  tSNE2 = reducedDims(sce.owen)$TSNE[,2],
                  colData(sce.owen))) + 
  geom_point(aes(tSNE1, tSNE2, colour = as.factor(gland_cells_k3)))
```