---
title: "Differential expression in stem cell compartment"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Reports/DE_stem_cells.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script performs differential expression between the stem cells of different tissues.

## Read in the data 

```{r data, message=FALSE}
library(scater)
library(scran)
library(edgeR)
library(ggplot2)
library(openxlsx)
library(RColorBrewer)
library(Rtsne)
library(umap)
source("../../Analysis/Functions/auxiliary.R")
sce <- readRDS("../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/All_corrected_sce.rds")
```

# OLFM4 cells

## Select cells

```{r}
# Select all OLFM4 cells
sce_olfm4 <- sce[,logcounts(sce)[rowData(sce)$Symbol == "OLFM4",] > 0]
plot(table(colData(sce_olfm4)$Tissue))
table(colData(sce_olfm4)$Tissue, colData(sce_olfm4)$Patient)
```

## Visualize cells

```{r}
# Tissue
ggplot(data.frame(tSNE1 = reducedDims(sce_olfm4)$TSNE[,1],
                  tSNE2 = reducedDims(sce_olfm4)$TSNE[,2],
                  tissue = colData(sce_olfm4)$Tissue)) +
  geom_point(aes(tSNE1, tSNE2, colour = tissue))

# Patient
ggplot(data.frame(tSNE1 = reducedDims(sce_olfm4)$TSNE[,1],
                  tSNE2 = reducedDims(sce_olfm4)$TSNE[,2],
                  patient = colData(sce_olfm4)$Patient)) +
  geom_point(aes(tSNE1, tSNE2, colour = patient))
```

# D2 vs BE

```{r}
sce.test <- sce_olfm4[,grepl("BE|D2", colData(sce_olfm4)$Tissue)]
sce.test <- sce.test[Matrix::rowMeans(logcounts(sce.test)) > 0.1,]
sce.test <- normalize(sce.test)
  
# Sum counts with each batch and group
mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce.test)$Patient, 
                                        colData(sce.test)$Tissue, sep = "_"))), 
              nrow = nrow(counts(sce.test)))
rownames(mat) <- rownames(counts(sce.test))
colnames(mat) <- unique(paste(colData(sce.test)$Patient, 
                                colData(sce.test)$Tissue, sep = "_"))
  
for(j in colnames(mat)){
  cur_patient <- unlist(strsplit(j, "_"))[1]
  cur_tissue <- unlist(strsplit(j, "_"))[2]
  mat[,j] <- Matrix::rowSums(counts(sce.test)[,colData(sce.test)$Patient == cur_patient &
                                                colData(sce.test)$Tissue == cur_tissue]) 
}
  
# Perform differential testing
y <- DGEList(counts=mat,
             group=sapply(colnames(mat), 
                          function(n){unlist(strsplit(n, "_"))[1]}))
y <- calcNormFactors(y)
design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
colnames(design) <- c("BE", "D2")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                contrast = makeContrasts(D2 - BE, levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table

# Save markers
cur_out <- list()
cur_out$BE <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.01,]
cur_out$BE$Genename <- rowData(sce)$Symbol[match(rownames(cur_out$BE), rowData(sce)$ID)]
cur_out$D2 <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.01,]
cur_out$D2$Genename <- rowData(sce)$Symbol[match(rownames(cur_out$D2), rowData(sce)$ID)]

write.xlsx(cur_out, "../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Marker_genes/DE/OLFM4_cells/BE_vs_D2.xlsx")
```

# Based on clustering

## Visualize clustering

```{r}
ggplot(data.frame(tSNE1 = reducedDims(sce)$TSNE[,1],
                  tSNE2 = reducedDims(sce)$TSNE[,2],
                  cluster = factor(colData(sce)$Tissue_cluster))) +
  geom_point(aes(tSNE1, tSNE2, colour = cluster)) + 
  scale_colour_manual(values = c(brewer.pal(9, "Set1"), 
                                 brewer.pal(8, "Set2"),
                                 brewer.pal(12, "Set3")))

ggplot(data.frame(tSNE1 = reducedDims(sce)$TSNE[,1],
                  tSNE2 = reducedDims(sce)$TSNE[,2],
                  cluster = factor(colData(sce)$Tissue))) +
  geom_point(aes(tSNE1, tSNE2, colour = cluster)) + 
  scale_colour_manual(values = c(brewer.pal(9, "Set1"), 
                                 brewer.pal(8, "Set2"),
                                 brewer.pal(12, "Set3")))

ggplot(data.frame(tSNE1 = reducedDims(sce)$TSNE[,1],
                  tSNE2 = reducedDims(sce)$TSNE[,2],
                  OLFM4 = logcounts(sce)[rowData(sce)$Symbol == "OLFM4",])) +
  geom_point(aes(tSNE1, tSNE2, colour = OLFM4)) + 
  scale_colour_viridis_c()

# Select possible stem cell clusters
sce.sub <- sce[,((colData(sce)$Tissue_cluster == 3 | 
                    colData(sce)$Tissue_cluster == 4) & 
                   colData(sce)$Tissue == "BE") |
                 ((colData(sce)$Tissue_cluster == 3) & 
                   colData(sce)$Tissue == "GC") |
                 (colData(sce)$Tissue_cluster == 4 & 
                   colData(sce)$Tissue == "D2")]

ggplot(data.frame(tSNE1 = reducedDims(sce.sub)$TSNE[,1],
                  tSNE2 = reducedDims(sce.sub)$TSNE[,2],
                  cluster = factor(colData(sce.sub)$Tissue_cluster))) +
  geom_point(aes(tSNE1, tSNE2, colour = cluster)) + 
  scale_colour_manual(values = c(brewer.pal(9, "Set1")))

ggplot(data.frame(tSNE1 = reducedDims(sce.sub)$TSNE[,1],
                  tSNE2 = reducedDims(sce.sub)$TSNE[,2],
                  cluster = factor(colData(sce.sub)$Tissue))) +
  geom_point(aes(tSNE1, tSNE2, colour = cluster)) + 
  scale_colour_manual(values = c(brewer.pal(9, "Set1")))

ggplot(data.frame(tSNE1 = reducedDims(sce.sub)$TSNE[,1],
                  tSNE2 = reducedDims(sce.sub)$TSNE[,2],
                  OLFM4 = logcounts(sce.sub)[rowData(sce.sub)$Symbol == "OLFM4",])) +
  geom_point(aes(tSNE1, tSNE2, colour = OLFM4)) + 
  scale_colour_viridis_c()

table(colData(sce.sub)$Tissue, colData(sce.sub)$Patient)
```

# Differential expression analysis

## GC vs BE

```{r}
sce.test <- sce.sub[,grepl("GC|BE", colData(sce.sub)$Tissue)]
sce.test <- sce.test[Matrix::rowMeans(logcounts(sce.test)) > 0.1,]
sce.test <- normalize(sce.test)
  
# Sum counts with each batch and group
mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce.test)$Patient, 
                                        colData(sce.test)$Tissue, sep = "_"))), 
              nrow = nrow(counts(sce.test)))
rownames(mat) <- rownames(counts(sce.test))
colnames(mat) <- unique(paste(colData(sce.test)$Patient, 
                                colData(sce.test)$Tissue, sep = "_"))
  
for(j in colnames(mat)){
  cur_patient <- unlist(strsplit(j, "_"))[1]
  cur_tissue <- unlist(strsplit(j, "_"))[2]
  mat[,j] <- Matrix::rowSums(counts(sce.test)[,colData(sce.test)$Patient == cur_patient &
                                                colData(sce.test)$Tissue == cur_tissue]) 
}
  
# Perform differential testing
y <- DGEList(counts=mat,
             group=sapply(colnames(mat), 
                          function(n){unlist(strsplit(n, "_"))[1]}))
y <- calcNormFactors(y)
design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
colnames(design) <- c("BE", "GC")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                contrast = makeContrasts(GC - BE, levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table

# Save markers
cur_out <- list()
cur_out$BE <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.01,]
cur_out$BE$Genename <- rowData(sce)$Symbol[match(rownames(cur_out$BE), rowData(sce)$ID)]
cur_out$GC <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.01,]
cur_out$GC$Genename <- rowData(sce)$Symbol[match(rownames(cur_out$GC), rowData(sce)$ID)]

write.xlsx(cur_out, "../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Marker_genes/DE/Cluster_based/BE_vs_GC.xlsx")
```

## BE vs D2

```{r}
sce.test <- sce.sub[,grepl("D2|BE", colData(sce.sub)$Tissue)]
sce.test <- sce.test[Matrix::rowMeans(logcounts(sce.test)) > 0.1,]
sce.test <- normalize(sce.test)
  
# Sum counts with each batch and group
mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce.test)$Patient, 
                                        colData(sce.test)$Tissue, sep = "_"))), 
              nrow = nrow(counts(sce.test)))
rownames(mat) <- rownames(counts(sce.test))
colnames(mat) <- unique(paste(colData(sce.test)$Patient, 
                                colData(sce.test)$Tissue, sep = "_"))
  
for(j in colnames(mat)){
  cur_patient <- unlist(strsplit(j, "_"))[1]
  cur_tissue <- unlist(strsplit(j, "_"))[2]
  mat[,j] <- Matrix::rowSums(counts(sce.test)[,colData(sce.test)$Patient == cur_patient &
                                                colData(sce.test)$Tissue == cur_tissue]) 
}
  
# Perform differential testing
y <- DGEList(counts=mat,
             group=sapply(colnames(mat), 
                          function(n){unlist(strsplit(n, "_"))[1]}))
y <- calcNormFactors(y)
design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
colnames(design) <- c("BE", "D2")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                contrast = makeContrasts(D2 - BE, levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table

# Save markers
cur_out <- list()
cur_out$BE <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.01,]
cur_out$BE$Genename <- rowData(sce)$Symbol[match(rownames(cur_out$BE), rowData(sce)$ID)]
cur_out$D2 <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.01,]
cur_out$D2$Genename <- rowData(sce)$Symbol[match(rownames(cur_out$D2), rowData(sce)$ID)]

write.xlsx(cur_out, "../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Marker_genes/DE/Cluster_based/BE_vs_D2.xlsx")
```

# Select cells based on Entropy measure

```{r}
# Select cells for analysis
cur_sce <- sce[,grepl("BE|SMG|GC|D2", colData(sce)$Tissue)]
ggplot(data.frame(tSNE1 = reducedDims(cur_sce)$TSNE[,1],
                  tSNE2 = reducedDims(cur_sce)$TSNE[,2],
                  tissue = colData(cur_sce)$Tissue)) +
  geom_point(aes(tSNE1, tSNE2, colour = tissue))

ggplot(data.frame(tSNE1 = reducedDims(cur_sce)$TSNE[,1],
                  tSNE2 = reducedDims(cur_sce)$TSNE[,2],
                  cluster = as.factor(colData(cur_sce)$Global_cluster))) +
  geom_point(aes(tSNE1, tSNE2, colour = cluster)) + 
  scale_color_manual(values = c(brewer.pal(n = 8, name = "Set1"),
                                brewer.pal(n = 8, name = "Set2"),
                                brewer.pal(n = 8, name = "Set3")))

cur_sce <- cur_sce[,colData(cur_sce)$Global_cluster %in% c(15, 13, 2, 1)]

# Perform new batch correction
sce.list <- split.sce(cur_sce, groups = unique(colData(cur_sce)$Sample), colData.name = "Sample")

# Order so that BE is first
sce.list <- sce.list[c(4,5,6,1,2,3,14,7,8,9,10,11,12,13)]
cur_sce <- do.call("cbind", sce.list)

corrected <- batch.correction(sce.list)

# Compute tSNE
set.seed(1234)
tsne <- Rtsne(t(corrected), pca = FALSE)
umap.cor <- umap(t(corrected))

# Visualize results
ggplot(data.frame(UMAP1 = umap.cor$layout[,1],
                  UMAP2 = umap.cor$layout[,2],
                  cluster = as.factor(colData(cur_sce)$Tissue))) +
  geom_point(aes(UMAP1, UMAP2, colour = cluster)) + 
  scale_color_manual(values = c(brewer.pal(n = 8, name = "Set1")))

ggplot(data.frame(tSNE1 = tsne$Y[,1],
                  tSNE2 = tsne$Y[,2],
                  cluster = as.factor(colData(cur_sce)$Tissue))) +
  geom_point(aes(tSNE1, tSNE2, colour = cluster)) + 
  scale_color_manual(values = c(brewer.pal(n = 8, name = "Set1")))

ggplot(data.frame(UMAP1 = umap.cor$layout[,1],
                  UMAP2 = umap.cor$layout[,2],
                  patient = as.factor(colData(cur_sce)$Patient))) +
  geom_point(aes(UMAP1, UMAP2, colour = patient)) + 
  scale_color_manual(values = c(brewer.pal(n = 12, name = "Set3")))

# Visualize only BE
ggplot(data.frame(UMAP1 = umap.cor$layout[colData(cur_sce)$Tissue == "BE",1],
                  UMAP2 = umap.cor$layout[colData(cur_sce)$Tissue == "BE",2],
                  patient = as.factor(colData(cur_sce[,colData(cur_sce)$Tissue == "BE"])$Patient))) +
  geom_point(aes(UMAP1, UMAP2, colour = patient)) + 
  scale_color_manual(values = c(brewer.pal(n = 12, name = "Set3")))

# Visualize only GC
ggplot(data.frame(UMAP1 = umap.cor$layout[colData(cur_sce)$Tissue == "GC",1],
                  UMAP2 = umap.cor$layout[colData(cur_sce)$Tissue == "GC",2],
                  patient = as.factor(colData(cur_sce[,colData(cur_sce)$Tissue == "GC"])$Patient))) +
  geom_point(aes(UMAP1, UMAP2, colour = patient)) + 
  scale_color_manual(values = c(brewer.pal(n = 12, name = "Set3")))

# Visualize only SMG
ggplot(data.frame(UMAP1 = umap.cor$layout[colData(cur_sce)$Tissue == "SMG",1],
                  UMAP2 = umap.cor$layout[colData(cur_sce)$Tissue == "SMG",2],
                  patient = as.factor(colData(cur_sce[,colData(cur_sce)$Tissue == "SMG"])$Patient))) +
  geom_point(aes(UMAP1, UMAP2, colour = patient)) + 
  scale_color_manual(values = c(brewer.pal(n = 12, name = "Set3")))

# Visualize only D2
ggplot(data.frame(UMAP1 = umap.cor$layout[colData(cur_sce)$Tissue == "D2",1],
                  UMAP2 = umap.cor$layout[colData(cur_sce)$Tissue == "D2",2],
                  patient = as.factor(colData(cur_sce[,colData(cur_sce)$Tissue == "D2"])$Patient))) +
  geom_point(aes(UMAP1, UMAP2, colour = patient)) + 
  scale_color_manual(values = c(brewer.pal(n = 12, name = "Set3")))

# Perform new clustering
set.seed(1234)
g <- buildSNNGraph(corrected, k = 15)
cluster <- igraph::cluster_louvain(g)$membership

ggplot(data.frame(UMAP1 = umap.cor$layout[,1],
                  UMAP2 = umap.cor$layout[,2],
                  cluster = as.factor(cluster))) +
  geom_point(aes(UMAP1, UMAP2, colour = cluster)) + 
  scale_color_manual(values = c(brewer.pal(n = 12, name = "Set3")))

ggplot(data.frame(UMAP1 = umap.cor$layout[,1],
                  UMAP2 = umap.cor$layout[,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "OLFM4",])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

table(cluster, colData(cur_sce)$Tissue, colData(cur_sce)$Patient)
```

# Calculate Entropy for each sample

```{r}
rm(sce)
entropy_vector <- vector(length = ncol(cur_sce))
names(entropy_vector) <- paste(cur_sce$Patient, cur_sce$Tissue, cur_sce$Barcode, sep = "_")

for(i in c("BE", "GC", "SMG", "D2")){
  sce_i <- cur_sce[,colData(cur_sce)$Tissue == i]
  sce_i <- sce_i[Matrix::rowSums(logcounts(sce_i)) > 0,]
  cur_counts <- 2^(logcounts(sce_i)) - 1 + 0.1
  probs   <- t(t(cur_counts)/Matrix::colSums(cur_counts))
  entropy <- -apply(probs*log(probs)/log(nrow(sce_i)),2,sum)

  entropy_vector[paste(sce_i$Patient, sce_i$Tissue, sce_i$Barcode, sep = "_")] <- entropy
}

ggplot(data.frame(UMAP1 = umap.cor$layout[,1],
                  UMAP2 = umap.cor$layout[,2],
                  entropy = entropy_vector)) +
  geom_point(aes(UMAP1, UMAP2, colour = entropy)) + 
  scale_color_viridis_c()
```

# DE between stem cells

# Cluster 1

```{r}
sce.stem <- cur_sce[,cluster == 1]
table(sce.stem$Tissue, sce.stem$Patient)

# Remove the 2 SMG cells
sce.stem <- sce.stem[,sce.stem$Tissue != "SMG"]

# Differential expression
# Sum counts with each batch and group
mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce.stem)$Patient, 
                                        colData(sce.stem)$Tissue, sep = "_"))), 
              nrow = nrow(counts(sce.stem)))
rownames(mat) <- rownames(counts(sce.stem))
colnames(mat) <- unique(paste(colData(sce.stem)$Patient, 
                                colData(sce.stem)$Tissue, sep = "_"))
  
for(j in colnames(mat)){
  cur_patient <- unlist(strsplit(j, "_"))[1]
  cur_tissue <- unlist(strsplit(j, "_"))[2]
  mat[,j] <- Matrix::rowSums(counts(sce.stem)[,colData(sce.stem)$Patient == cur_patient &
                                                colData(sce.stem)$Tissue == cur_tissue]) 
}
  
# Perform differential testing
y <- DGEList(counts=mat,
             group=sapply(colnames(mat), 
                          function(n){unlist(strsplit(n, "_"))[1]}))
y <- calcNormFactors(y)
design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
colnames(design) <- c("BE", "GC")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                contrast = makeContrasts(GC - BE, levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table

# Save markers
cur_out <- list()
cur_out$BE <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.1,]
cur_out$BE$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$BE), rowData(sce.stem)$ID)]
cur_out$GC <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.1,]
cur_out$GC$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$GC), rowData(sce.stem)$ID)]

# Visualize gene expression
ggplot(data.frame(UMAP1 = umap.cor$layout[cluster == 1,1],
                  UMAP2 = umap.cor$layout[cluster == 1,2],
                  tissue = cur_sce$Tissue[cluster == 1])) +
  geom_point(aes(UMAP1, UMAP2, colour = tissue))

ggplot(data.frame(UMAP1 = umap.cor$layout[cluster == 1,1],
                  UMAP2 = umap.cor$layout[cluster == 1,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "S100A10",
                                            cluster == 1])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

ggplot(data.frame(UMAP1 = umap.cor$layout[cluster == 1,1],
                  UMAP2 = umap.cor$layout[cluster == 1,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "GHRL",
                                            cluster == 1])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

# Save Excel 
write.xlsx(cur_out, "../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Marker_genes/DE/Stem_cells/DE_GC_vs_BE_stem_cells_cluster1.xlsx")
```

# Cluster 6

```{r}
sce.stem <- cur_sce[,cluster == 6]
table(sce.stem$Tissue, sce.stem$Patient)

# Remove the 2 SMG cells
sce.stem <- sce.stem[,sce.stem$Tissue != "SMG" &
                       sce.stem$Tissue != "GC"]

# Differential expression
# Sum counts with each batch and group
mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce.stem)$Patient, 
                                        colData(sce.stem)$Tissue, sep = "_"))), 
              nrow = nrow(counts(sce.stem)))
rownames(mat) <- rownames(counts(sce.stem))
colnames(mat) <- unique(paste(colData(sce.stem)$Patient, 
                                colData(sce.stem)$Tissue, sep = "_"))
  
for(j in colnames(mat)){
  cur_patient <- unlist(strsplit(j, "_"))[1]
  cur_tissue <- unlist(strsplit(j, "_"))[2]
  mat[,j] <- Matrix::rowSums(counts(sce.stem)[,colData(sce.stem)$Patient == cur_patient &
                                                colData(sce.stem)$Tissue == cur_tissue]) 
}
  
# Perform differential testing
y <- DGEList(counts=mat,
             group=sapply(colnames(mat), 
                          function(n){unlist(strsplit(n, "_"))[1]}))
y <- calcNormFactors(y)
design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
colnames(design) <- c("BE", "D2")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                contrast = makeContrasts(D2 - BE, levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table

# Save markers
cur_out <- list()
cur_out$BE <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.1,]
cur_out$BE$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$BE), rowData(sce.stem)$ID)]
cur_out$D2 <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.1,]
cur_out$D2$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$D2), rowData(sce.stem)$ID)]

# Visualize gene expression
ggplot(data.frame(UMAP1 = umap.cor$layout[cluster == 6,1],
                  UMAP2 = umap.cor$layout[cluster == 6,2],
                  tissue = cur_sce$Tissue[cluster == 6])) +
  geom_point(aes(UMAP1, UMAP2, colour = tissue))

ggplot(data.frame(UMAP1 = umap.cor$layout[cluster == 6,1],
                  UMAP2 = umap.cor$layout[cluster == 6,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "REG1A",
                                            cluster == 6])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

ggplot(data.frame(UMAP1 = umap.cor$layout[cluster == 6,1],
                  UMAP2 = umap.cor$layout[cluster == 6,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "TFF1",
                                            cluster == 6])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

# Save Excel 
write.xlsx(cur_out, "../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Marker_genes/DE/Stem_cells/DE_D2_vs_BE_stem_cells_cluster6.xlsx")
```

# BE Cluster 6 vs GC Cluster 1

```{r}
sce.stem <- cur_sce[,(cur_sce$Tissue == "GC" & cluster == 1) |
                      (cur_sce$Tissue == "BE" & cluster == 6)]
table(sce.stem$Tissue, sce.stem$Patient)

# Differential expression
# Sum counts with each batch and group
mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce.stem)$Patient, 
                                        colData(sce.stem)$Tissue, sep = "_"))), 
              nrow = nrow(counts(sce.stem)))
rownames(mat) <- rownames(counts(sce.stem))
colnames(mat) <- unique(paste(colData(sce.stem)$Patient, 
                                colData(sce.stem)$Tissue, sep = "_"))
  
for(j in colnames(mat)){
  cur_patient <- unlist(strsplit(j, "_"))[1]
  cur_tissue <- unlist(strsplit(j, "_"))[2]
  mat[,j] <- Matrix::rowSums(counts(sce.stem)[,colData(sce.stem)$Patient == cur_patient &
                                                colData(sce.stem)$Tissue == cur_tissue]) 
}
  
# Perform differential testing
y <- DGEList(counts=mat,
             group=sapply(colnames(mat), 
                          function(n){unlist(strsplit(n, "_"))[1]}))
y <- calcNormFactors(y)
design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
colnames(design) <- c("BE", "GC")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                contrast = makeContrasts(GC - BE, levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table

# Save markers
cur_out <- list()
cur_out$BE <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.1,]
cur_out$BE$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$BE), rowData(sce.stem)$ID)]
cur_out$GC <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.1,]
cur_out$GC$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$GC), rowData(sce.stem)$ID)]

# Visualize gene expression
cur_cluster <- (cur_sce$Tissue == "GC" & cluster == 1) |
                      (cur_sce$Tissue == "BE" & cluster == 6)
ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  tissue = cur_sce$Tissue[cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = tissue))

ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  patient = cur_sce$Patient[cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = patient))

ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "REG4",
                                            cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "CXCL17",
                                            cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

# Save Excel 
write.xlsx(cur_out, "../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Marker_genes/DE/Stem_cells/DE_GC_cluster1_vs_BE_cluster6.xlsx")
```

# D2 Cluster 6 vs GC Cluster 1

```{r}
sce.stem <- cur_sce[,(cur_sce$Tissue == "GC" & cluster == 1) |
                      (cur_sce$Tissue == "D2" & cluster == 6)]
table(sce.stem$Tissue, sce.stem$Patient)

# Differential expression
# Sum counts with each batch and group
mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce.stem)$Patient, 
                                        colData(sce.stem)$Tissue, sep = "_"))), 
              nrow = nrow(counts(sce.stem)))
rownames(mat) <- rownames(counts(sce.stem))
colnames(mat) <- unique(paste(colData(sce.stem)$Patient, 
                                colData(sce.stem)$Tissue, sep = "_"))
  
for(j in colnames(mat)){
  cur_patient <- unlist(strsplit(j, "_"))[1]
  cur_tissue <- unlist(strsplit(j, "_"))[2]
  mat[,j] <- Matrix::rowSums(counts(sce.stem)[,colData(sce.stem)$Patient == cur_patient &
                                                colData(sce.stem)$Tissue == cur_tissue]) 
}
  
# Perform differential testing
y <- DGEList(counts=mat,
             group=sapply(colnames(mat), 
                          function(n){unlist(strsplit(n, "_"))[1]}))
y <- calcNormFactors(y)
design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
colnames(design) <- c("D2", "GC")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                contrast = makeContrasts(GC - D2, levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table

# Save markers
cur_out <- list()
cur_out$D2 <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.1,]
cur_out$D2$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$D2), rowData(sce.stem)$ID)]
cur_out$GC <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.1,]
cur_out$GC$Genename <- rowData(sce.stem)$Symbol[match(rownames(cur_out$GC), rowData(sce.stem)$ID)]

# Visualize gene expression
cur_cluster <- (cur_sce$Tissue == "GC" & cluster == 1) |
                      (cur_sce$Tissue == "D2" & cluster == 6)
ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  tissue = cur_sce$Tissue[cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = tissue))

ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  patient = cur_sce$Patient[cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = patient))

ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "OLFM4",
                                            cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

ggplot(data.frame(UMAP1 = umap.cor$layout[cur_cluster,1],
                  UMAP2 = umap.cor$layout[cur_cluster,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "TFF1",
                                            cur_cluster])) +
  geom_point(aes(UMAP1, UMAP2, colour = gene)) + 
  scale_color_viridis_c()

# Save Excel 
write.xlsx(cur_out, "../../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Marker_genes/DE/Stem_cells/DE_GC_cluster1_vs_D2_cluster6.xlsx")
```
