---
title: "Figure 2: Similarity analysis between BE and other tissues"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Figures/Fig_2.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script focuses on the similarity analysis between BE and other tissues: SCJ, GOJ, GC, SMG and D2.

# Read in the data

```{r libraries, message=FALSE}
library(scran)
library(scater)
library(DropletUtils)
library(openxlsx)
library(Rtsne)
library(pheatmap)
library(cowplot)
library(RColorBrewer)
library(edgeR)
library(ape)
source("../Analysis/Functions/auxiliary.R")
sce <- readRDS("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/All_corrected_sce.rds")
```

# Visualize GC, GOJ, BE and SCJ samples

Here we perform differential expression (DE) analysis between the epithelial tissues of the GC, GOJ, SCJ, BE and NE samples.

```{r}
# Select tissues to test
#cur_corrected <- metadata(sce)$corrected[,sce$include & sce$Tissue %in% c("NE", "GC", "BE", "GOJ", "SCJ")]
cur_sce <- sce[,sce$include & sce$Tissue %in% c("NE", "GC", "BE", "GOJ", "SCJ")]
cur_sce <- normalize(cur_sce)

# Perform batch correction
sce.list <- split.sce(cur_sce, unique(cur_sce$Sample), colData.name = "Sample")
sce.list <- sce.list[c(which(grepl("NE", names(sce.list))),
                       which(grepl("GOJ", names(sce.list))),
                       which(grepl("SCJ", names(sce.list))),
                       which(grepl("GC", names(sce.list))),
                       which(grepl("BE", names(sce.list))))]
corrected <- batch.correction(sce.list)
cur_sce <- do.call("cbind", sce.list)

# Compute new tSNE
set.seed(11111)
tsne <- Rtsne(t(corrected), pca = FALSE, perplexity = 100)
reducedDims(cur_sce)$TSNE <- tsne$Y

# Visualize tSNEs
# Tissue
p.all.cells <- ggplot(data.frame(tsne1 = reducedDims(cur_sce)$TSNE[,1],
                  tsne2 = reducedDims(cur_sce)$TSNE[,2],
                  tissue = colData(cur_sce)$Tissue)) + 
  geom_point(aes(tsne1, tsne2), colour = "black", size = 2) + 
  geom_point(aes(tsne1, tsne2, colour = tissue)) + 
  scale_color_manual(values = metadata(cur_sce)$colour_vector) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "grey"))

# Tissue typse
p.all.cells.tissue_type <- ggplot(data.frame(tsne1 = reducedDims(cur_sce)$TSNE[,1],
                  tsne2 = reducedDims(cur_sce)$TSNE[,2],
                  tissue = colData(cur_sce)$tissue_type)) +
  geom_point(aes(tsne1, tsne2), colour = "black", size = 2) + 
  geom_point(aes(tsne1, tsne2, colour = tissue)) + 
  scale_color_manual(values = c("dark blue", "dark red")) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "grey"))

# Cell type
colour_vector <- vector(length = length(unique(cur_sce$cell_type)))
names(colour_vector) <- unique(cur_sce$cell_type)
colour_vector["Unknown"] <- "grey20"
colour_vector["Suprabasal"] <- colorRampPalette(c("white", "dark red"))(10)[10]
colour_vector["Basal"] <- colorRampPalette(c("white", "dark red"))(10)[4]
colour_vector["Intermediate"] <- colorRampPalette(c("white", "dark red"))(10)[7]
colour_vector["Dividing"] <- colorRampPalette(c("white", "dark red"))(10)[1]
colour_vector["Foveolar_differentiated"] <- colorRampPalette(c("white", "#4DBBD5FF"))(10)[10]
colour_vector["Endocrine"] <- colorRampPalette(c("white", "dark blue"))(10)[10]
colour_vector["Undifferentiated"] <- colorRampPalette(c("white", "#4DBBD5FF"))(10)[3]
colour_vector["Foveolar_Intermediate"] <- colorRampPalette(c("white", "#4DBBD5FF"))(10)[6]
colour_vector["Chief"] <- colorRampPalette(c("white", "dark blue"))(10)[6]
colour_vector["Parietal"] <- colorRampPalette(c("white", "dark blue"))(10)[3]
colour_vector["Foveolar_Intermediate/goblet"] <- "black"
colour_vector["Endocrine/Undifferentiated"] <- "grey60"
colour_vector["Goblet"] <- colorRampPalette(c("white", "seagreen4"))(10)[10]

p.all.cells.cell_type <- ggplot(data.frame(tsne1 = reducedDims(cur_sce)$TSNE[,1],
                  tsne2 = reducedDims(cur_sce)$TSNE[,2],
                  tissue = colData(cur_sce)$cell_type)) +
  geom_point(aes(tsne1, tsne2), colour = "black", size = 2) + 
  geom_point(aes(tsne1, tsne2, colour = tissue)) + 
  scale_color_manual(values = colour_vector) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "grey"))

# Visualize patient
# Colouring for patient
colour_vector <- vector(length = length(unique(cur_sce$Patient)))
names(colour_vector) <- unique(cur_sce$Patient)
colour_vector <- brewer.pal(7, "Accent")

p.all.cells.patient <- ggplot(data.frame(tsne1 = reducedDims(cur_sce)$TSNE[,1],
                  tsne2 = reducedDims(cur_sce)$TSNE[,2],
                  tissue = colData(cur_sce)$Patient)) +
  geom_point(aes(tsne1, tsne2), colour = "black", size = 2) + 
  geom_point(aes(tsne1, tsne2, colour = tissue)) + 
  scale_color_manual(values = colour_vector) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "grey"))

# Save tSNEs
final <- plot_grid(p.all.cells, p.all.cells.tissue_type, p.all.cells.cell_type, p.all.cells.patient,
                   ncol = 2, nrow = 2)
ggsave("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Figures/Figure_2/GC_NE_tSNEs.pdf", width = 15, height = 15)
```

# Differential expression analysis

Here we berform differential expression analysis between NE and squamous cells of GOJ, GC and columnar cells of GC, NE and squamous cells of SCJ, BE and columnar cells of SCJ.
For this, we will sum the raw counts for each gene across all cells per tissue and per patient. 
We then use edgeR on these pseudo-bulk samples to test for changes in mean expression.

```{r}
# Define conditions to test
# First two slots are tissues and last slot is cell-types to compare
cond <- list(c("NE", "GOJ", "Basal-Intermediate-Suprabasal"),
             c("GC", "GOJ", "Foveolar_Intermediate-Foveolar_differentiated-Endocrine-Undifferentiated-Chief"),
             c("BE", "SCJ", "Foveolar_differentiated-Foveolar_Intermediate-Undifferentiated-Dividing"),
             c("NE", "SCJ", "Basal-Intermediate-Suprabasal"))

# Initialize list to store results
out <- list()

for(i in 1:length(cond)){
  tissue1 <- cond[[i]][1]
  tissue2 <- cond[[i]][2]
  cell_types <- unlist(strsplit(cond[[i]][3], "-"))
  
  # Select cells
  sce_to_test <- cur_sce[,(cur_sce$Tissue == tissue1 | cur_sce$Tissue == tissue2) &
                           cur_sce$cell_type %in% cell_types]
  
  # Sum counts with each batch and group
  mat <- matrix(data = NA, ncol = length(unique(paste(colData(sce_to_test)$Patient, 
                                          colData(sce_to_test)$Tissue, sep = "_"))), 
               nrow = nrow(counts(sce_to_test)))
  rownames(mat) <- rownames(counts(sce_to_test))
  colnames(mat) <- unique(paste(colData(sce_to_test)$Patient, 
                                  colData(sce_to_test)$Tissue, sep = "_"))
  
  for(j in colnames(mat)){
    cur_patient <- unlist(strsplit(j, "_"))[1]
    cur_tissue <- unlist(strsplit(j, "_"))[2]
    mat[,j] <- Matrix::rowSums(counts(sce_to_test)[,colData(sce_to_test)$Patient == cur_patient &
                                                  colData(sce_to_test)$Tissue == cur_tissue]) 
  }
  
  # Remove lowly expressed genes after scale normalizing, rpm - reads per million
  rpm <- t(t(mat)/(colSums(mat)/1000000))
  mat <- mat[rowMeans(rpm) > 1,]
  
  # Perform differential testing
  y <- DGEList(counts=mat,
               group=sapply(colnames(mat), 
                           function(n){unlist(strsplit(n, "_"))[1]}))
  y <- calcNormFactors(y)
  design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[2]}))
  colnames(design) <- substring(colnames(design), regexpr("})", colnames(design)) + 2)
  y <- estimateDisp(y,design)
  
  fit <- glmQLFit(y,design, robust = TRUE)
  qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
    contrast = eval(parse(text = paste("makeContrasts(", tissue1,  " - ", 
                                       tissue2, ", levels = design)", sep = ""))))
  cur_markers <- topTags(qlf, n = nrow(qlf$table))$table
  
  # Print the number of genes being differentially expressed
  print(sum(cur_markers$FDR < 0.1)/nrow(mat))
  
  # Save markers
  cur_out <- list()
  cur_out[[tissue2]] <- cur_markers[cur_markers$logFC < 0 & cur_markers$FDR < 0.1,]
  cur_out[[tissue2]]$Genename <- rowData(sce_to_test)$Symbol[match(rownames(cur_out[[tissue2]]),
                                                                   rowData(sce_to_test)$ID)]
  cur_out[[tissue1]] <- cur_markers[cur_markers$logFC > 0 & cur_markers$FDR < 0.1,]
  cur_out[[tissue1]]$Genename <- rowData(sce_to_test)$Symbol[match(rownames(cur_out[[tissue1]]),
                                                                   rowData(sce_to_test)$ID)]
  
  # Save in output list
  out[[paste(tissue1, "vs", tissue2, sep = "_")]] <- cur_out
}

```

# Visualize KRT7 expression

Here, we visualize the expression of KRT7 in the tissues analysed above.

```{r}
KRT7 <- data.frame(KRT7 = logcounts(cur_sce)["ENSG00000135480",],
                   tissue = factor(cur_sce$Tissue, levels = c("NE", "GOJ", "GC", "SCJ", "BE")),
                   tissue_type = factor(cur_sce$tissue_type, levels = c("Squamous", "Columnar")))

p.KRT7 <- ggplot(KRT7) +
  geom_boxplot(aes(tissue, KRT7, fill = tissue_type, colour = tissue_type)) + 
  scale_colour_manual(values = c("dark red", "dark blue")) +
  scale_fill_manual(values = c(adjustcolor("dark red", alpha.f = 0.5), 
                               adjustcolor("dark blue", alpha.f = 0.5)))

ggsave("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Figures/Figure_2/KRT7_boxplots.pdf", width = 7, height = 4)
```

# Plot circular dendrogram of all clusters

To estimate the similarity between all cell types, we compute an averaged expression profile for each cell-type and perform hierarchical cluster analysis.

```{r}
# We will perform this analysis on the corrected and uncorrected counts (uncorrected in supplements)
clusters <- paste(colData(sce)$Tissue, colData(sce)$Tissue_cluster, sep = "_")
mat <- matrix(data = NA, ncol = length(unique(clusters)), nrow = nrow(metadata(sce)$corrected))
colnames(mat) <- unique(clusters)

for(i in unique(clusters)){
  mat[,i] <- rowMeans(metadata(sce)$corrected[,clusters == i])
}

# Rename the matrix to contain the actual cell-type labels
for(i in 1:ncol(mat)){
  colnames(mat)[i] <- unique(sce$cell_type[sce$Tissue == sub("_[0-9]*$", "", colnames(mat)[i]) &
                                 sce$Tissue_cluster == sub("^[A-Z2]*_", "", colnames(mat)[i])])
}

# Calculate euclidean distance on corrected counts
dend <- hclust(dist(t(mat), method = "euclidean"), method = "ward.D2")
pdf("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Figures/Figure_2/Tree_corrected_counts.pdf", width = 10, height = 10)
plot(as.phylo(dend), type = "fan", 
     tip.color = metadata(sce)$colour_vector[sub("_[0-9]*$", "", unique(clusters))])
dev.off()

# Plot first two components
pca.p <- ggplot(data.frame(Component1 = mat[1,],
                  Component2 = mat[2,],
                  tissue_type = sub("_[0-9]*$", "", unique(clusters)),
                  cell_type = colnames(mat))) +
  geom_point(aes(Component1, Component2, colour = tissue_type)) +
  scale_colour_manual(values = metadata(sce)$colour_vector) + 
  geom_text(aes(Component1, Component2, label = cell_type))

ggsave("../../../Dropbox (Cambridge University)/Oesophagus_single_cell/Results/Figures/Figure_2/PCA_cell_types_corrected.pdf", pca.p, width = 10, height = 10)
```

