---
title: "Figure 1: overview on experimental approach and gland cells"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script visualizes the full dataset as well as highlighting the sub-populations in the gland cells.

# Read in the data

```{r libraries, message=FALSE}
library(scran)
library(scater)
library(DropletUtils)
library(openxlsx)
library(Rtsne)
library(pheatmap)
library(viridis)
source("../Analysis/Functions/auxiliary.R")

# Read in the normalized and batch-corrected reads
sce <- readRDS("../../../Dropbox (Personal)/Oesophagus_single_cell/All_corrected_sce.rds")
```

# Visualize gland cells

To visualize the gland cells, we perform batch correction across all patients and colour cells based on the clustering performed in the Tissue_correction script.

```{r}
cur_sce <- sce[,sce$Tissue == "SMG"]

# Split the SCE object into individual patients
sce.list <- split.sce(cur_sce, unique(cur_sce$Patient), colData.name = "Patient")

# Perform batch correction
corrected <- batch.correction(sce.list)

# Merge sce objects
cur_sce <- do.call("cbind", sce.list)

# Compute tSNE for gland cells
set.seed(111)
smg_tsne <- Rtsne(t(corrected), pca = FALSE)

# Create colour vecotr for glands
colour_vector <-  vector(length = length(unique(cur_sce$cell_type)))
names(colour_vector) <- unique(cur_sce$cell_type)
colour_vector["Fibroblast"] <- "grey40"
colour_vector["Doublets.unknown"] <- "white"
colour_vector["Mucous"] <- "saddlebrown"
colour_vector["Serous"] <- "burlywood3"
colour_vector["Endothelial"] <- "grey60"
colour_vector["Nonepithelial"] <- "grey80"
colour_vector["Duct_intercalating"] <- "burlywood4"
colour_vector["Squamous_Esophagus"] <- "black"
colour_vector["Myo-epithelial"] <- "brown"
colour_vector["Dendritic"] <- "grey20"

p.smg <- ggplot(data.frame(tsne1 = smg_tsne$Y[,1],
                  tsne2 = smg_tsne$Y[,2],
                  cell_type = as.factor(cur_sce$cell_type))) +
  geom_point(aes(tsne1, tsne2), colour = "black", size = 2) + 
  geom_point(aes(tsne1, tsne2, colour = cell_type)) + 
  scale_color_manual(values = colour_vector) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "grey"))

ggsave("../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_1/SMG_cells.pdf", 
       p.smg, 
       width = 8, height = 7)
```

# Find SMG marker genes

Here, we use edgeR to perform pairwise differential (DE) analysis between all SMG cell types.
The pairwise comparisons are then combined to find cell-type specific marker genes.

```{r}
# Remove non-expressed genes
cur_sce <- cur_sce[Matrix::rowSums(counts(cur_sce)) > 0,]

# Perform DE to find marker genes
DE.genes <- multi.DE(sce = cur_sce, 
                     conditions = cur_sce$cell_type,
                     covariate = cur_sce$Patient,
                     lfc = 0.5,
                     FDR = 0.1)

# We first normalize gene expression across all selected cells
cur_sce <- computeSumFactors(cur_sce, clusters=cur_sce$cell_type)
cur_sce <- normalize(cur_sce, return_log = TRUE)

# Alternative DE
markers <- marker.detection(cur_sce, clusters = cur_sce$cell_type)

# compare to scran
match(DE.genes$Fibroblast$Symbol, markers$Fibroblast$GeneName)
match(DE.genes$Duct_intercalating$Symbol, markers$Duct_intercalating$GeneName)
match(DE.genes$Mucous$Symbol, markers$Mucous$GeneName)
match(DE.genes$Doublets.unknown$Symbol, markers$Doublets.unknown$GeneName)
match(DE.genes$Endothelial$Symbol, markers$Endothelial$GeneName)
match(DE.genes$Nonepithelial$Symbol, markers$Nonepithelial$GeneName)
match(DE.genes$Squamous_Esophagus$Symbol, markers$Squamous_Esophagus$GeneName)
match(DE.genes$Serous$Symbol, markers$Serous$GeneName)
match(DE.genes$Myo_epithelial$Symbol, markers$`Myo-epithelial`$GeneName)
match(DE.genes$Dendritic$Symbol, markers$Dendritic$GeneName)

# Visualize marker genes
cur_genes <- rownames(DE.genes$Fibroblast)
cur_counts <- logcounts(cur_sce)[cur_genes,order(cur_sce$cell_type)]
colnames(cur_counts) <- paste(cur_sce$Barcode, cur_sce$Patient, sep = "_")
pheatmap(as.matrix(cur_counts), scale = "row", 
         color = colorRampPalette(c("dark blue", "dark blue", "dark blue", "white", "dark red", "dark red", "dark red"))(100), cluster_rows = FALSE,
         cluster_cols = FALSE, 
         annotation_col = data.frame(row.names = colnames(cur_counts),
           cell_type = cur_sce$cell_type[order(cur_sce$cell_type)],
           patient = cur_sce$Patient[order(cur_sce$cell_type)]), 
         show_colnames = FALSE, show_rownames = FALSE)

# Visualize gene expression for control
ggplot(data.frame(tsne1 = smg_tsne$Y[,1],
                  tsne2 = smg_tsne$Y[,2],
                  gene = logcounts(cur_sce)[rowData(cur_sce)$Symbol == "CASC8",])) + 
  geom_point(aes(tsne1, tsne2, colour = gene)) +
  scale_colour_viridis()

# Save as table
write.xlsx(DE.genes, file = "../../../Dropbox (Personal)/Origin_of_BE_draft/Figures/Supplemental_tables/Table_S2.xlsx")
```

# Heatmap visualizing SMG marker genes

Here, we display marker genes selected from the list of cluster-specific marker genes as generated by the Tissue_correction script.

```{r}
genes <- c("CDH1", "TFF3", "MUC5B", "AGR2", "KRT8", "SLPI", 
           "MMP7", "KRT7", "SOX9", "KRT23", "ACTA2", "KRT5",
           "MKI67", "OLFM4")

# Remove somatic cells
cur_corrected <- corrected[,cur_sce$cell_type %in% 
                     c("Mucous", "Serous", "Duct_intercalating", "Myo-epithelial")]
sel_sce <- cur_sce[,cur_sce$cell_type %in% 
                     c("Mucous", "Serous", "Duct_intercalating", "Myo-epithelial")]
sel_sce <- normalize(sel_sce)
for.heatmap <- logcounts(sel_sce)[match(genes, rowData(sel_sce)$Symbol),]
colnames(for.heatmap) <- paste(sel_sce$Barcode, sel_sce$Patient, sep = "_")

# Calculate euclidean distance using the factors after batch correction
euc.dist <- dist(t(cur_corrected), method = "euclidean")

# Colouring for patient
patient_vector <- c("dark red", "dark blue", "dark green")
names(patient_vector) <- c("Patient10", "Patient11", "Patient13")

dev.off()
pdf(file = "../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_1/SMG_heatmap_lognorm.pdf", height = 7, width = 8)
pheatmap(for.heatmap, cluster_rows = FALSE, color = viridis(100), 
         labels_row = genes, show_colnames = FALSE, clustering_distance_cols = euc.dist,
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                                     cell_type = sel_sce$cell_type,
                                     patient = sel_sce$Patient), 
         annotation_colors = list(cell_type = colour_vector,
                                  patient = patient_vector), 
         clustering_method = "ward.D2")
dev.off()
```

