---
title: "Figure 3: differential testing between BE and GC"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Here, we will find genes that are differentially expressed between GC and BE for stem-like cells and differenitatied cells.

# Read in the data

```{r libraries, message=FALSE}
library(scran)
library(scater)
library(DropletUtils)
library(openxlsx)
library(Rtsne)
library(pheatmap)
library(cowplot)
library(RColorBrewer)
library(edgeR)
library(ape)
library(viridis)
library(umap)
library(reshape2)
source("../Analysis/Functions/auxiliary.R")
sce <- readRDS("../../../Dropbox (Personal)/Oesophagus_single_cell/All_corrected_sce.rds")
```

# Perform dimensionality reduction for GC and BE together

To find genes that are cell-type specific and differentially expressed between GC and BE, we focus the analysis only on these 2 tissues.

```{r}
# Select tissues 
cur_sce <- sce[,sce$include & sce$Tissue %in% c("GC", "BE")]

# Deselect cell types that do not match between tissues
cur_sce <- cur_sce[,!(cur_sce$cell_type %in% c("Chief", "Endocrine", "Goblet", "Parietal"))]

# Remove genes that are not expressed
cur_sce <- cur_sce[Matrix::rowSums(counts(cur_sce)) > 0,]

# Perform batch correction
sce.list <- split.sce(cur_sce, unique(cur_sce$Sample), colData.name = "Sample")
sce.list <- sce.list[c(which(grepl("GC", names(sce.list))),
                       which(grepl("BE", names(sce.list))))]
corrected <- batch.correction(sce.list)
cur_sce <- do.call("cbind", sce.list)

# Dimensionality reduction
UMAP <- umap(t(corrected))

UMAP_BE_GC.tissue <- ggplot(data.frame(UMAP1 = UMAP$layout[,1],
                  UMAP2 = UMAP$layout[,2],
                  tissue = cur_sce$Tissue)) +
  geom_point(aes(UMAP1, UMAP2, colour = tissue)) +
  scale_colour_manual(values = metadata(cur_sce)$colour_vector) + theme_minimal()

set.seed(1234)
tsne <- Rtsne(t(corrected), pca = FALSE)

tsne_BE_GC.tissue <- ggplot(data.frame(tsne1 = tsne$Y[,1],
                  tsne2 = tsne$Y[,2],
                  tissue = cur_sce$Tissue)) +
  geom_point(aes(tsne1, tsne2, colour = tissue)) +
  scale_colour_manual(values = metadata(cur_sce)$colour_vector)

# Colour vector for cell-types
colour_vector <- vector(length = length(unique(cur_sce$cell_type)))
names(colour_vector) <- unique(cur_sce$cell_type)
colour_vector["Dividing"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[3]
colour_vector["Foveolar_differentiated"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[10]
colour_vector["Foveolar_Intermediate"] <- colorRampPalette(c("white", "#4DBBD5FF"))(10)[6]
colour_vector["Undifferentiated"] <- colorRampPalette(c("white", "#3C5488FF"))(10)[1]

UMAP_BE_GC.cell_type <- ggplot(data.frame(UMAP1 = UMAP$layout[,1],
                  UMAP2 = UMAP$layout[,2],
                  tissue = cur_sce$cell_type)) +
  geom_point(aes(UMAP1, UMAP2), colour = "black", size = 2) +
  geom_point(aes(UMAP1, UMAP2, colour = tissue)) +
  scale_colour_manual(values = colour_vector) + theme_minimal()

UMAP_BE_GC.patient <- ggplot(data.frame(UMAP1 = UMAP$layout[,1],
                  UMAP2 = UMAP$layout[,2],
                  tissue = cur_sce$Patient)) +
  geom_point(aes(UMAP1, UMAP2), colour = "black", size = 2) +
  geom_point(aes(UMAP1, UMAP2, colour = tissue)) + theme_minimal()

tsne_BE_GC.cell_type <- ggplot(data.frame(tsne1 = tsne$Y[,1],
                  tsne2 = tsne$Y[,2],
                  tissue = cur_sce$cell_type)) +
  geom_point(aes(tsne1, tsne2), colour = "black", size = 2) +
  geom_point(aes(tsne1, tsne2, colour = tissue)) +
  scale_colour_manual(values = colour_vector) + theme_minimal()

# Save figures
ggsave("../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/BE_GC_tissue.pdf", UMAP_BE_GC.tissue, width = 7, height = 5)
ggsave("../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/BE_GC_cell_type.pdf", UMAP_BE_GC.cell_type, width = 7, height = 5)
```

# Differential expression testing between cell types

Here, we perform DE between progenitor and differentiated cell types and across tissues.

```{r}
# List to save result
out <- list()

# DE within GC tissue
sce_test <- cur_sce[,cur_sce$Tissue == "GC" & 
                      cur_sce$cell_type %in% c("Undifferentiated", "Foveolar_differentiated")]

# We set the FDR to 1 to retain all genes
GC_Undiff_vs_Diff <- DE.edgeR(sce_test, conditions = sce_test$cell_type,
                              covariate = sce_test$Patient, lfc = 0.5,
                              FDR = 1)
GC_Undiff_vs_Diff.out <- rbind(GC_Undiff_vs_Diff$Undifferentiated,
                               GC_Undiff_vs_Diff$Foveolar_differentiated[
              seq(nrow(GC_Undiff_vs_Diff$Foveolar_differentiated), 1),])

GC_Undiff_vs_Diff.out$ID <- rownames(GC_Undiff_vs_Diff.out)

GC_Undiff_vs_Diff.out$specificity <- NA
GC_Undiff_vs_Diff.out$specificity[GC_Undiff_vs_Diff.out$logFC < 0 & GC_Undiff_vs_Diff.out$FDR <= 0.1] <- "Undifferentiated"
GC_Undiff_vs_Diff.out$specificity[GC_Undiff_vs_Diff.out$logFC > 0 & GC_Undiff_vs_Diff.out$FDR <= 0.1] <- "Foveolar_differentiated"

out$GC_Undiff_vs_Diff <- GC_Undiff_vs_Diff.out

# DE within BE tissue
sce_test <- cur_sce[,cur_sce$Tissue == "BE" & cur_sce$cell_type %in% c("Undifferentiated", "Foveolar_differentiated")]
BE_Undiff_vs_Diff <- DE.edgeR(sce_test, conditions = sce_test$cell_type,
                              covariate = sce_test$Patient, lfc = 0.5,
                              FDR = 1)

BE_Undiff_vs_Diff.out <- rbind(BE_Undiff_vs_Diff$Undifferentiated,
                               BE_Undiff_vs_Diff$Foveolar_differentiated[
              seq(nrow(BE_Undiff_vs_Diff$Foveolar_differentiated), 1),])

BE_Undiff_vs_Diff.out$ID <- rownames(BE_Undiff_vs_Diff.out)

BE_Undiff_vs_Diff.out$specificity <- NA
BE_Undiff_vs_Diff.out$specificity[BE_Undiff_vs_Diff.out$logFC < 0 & BE_Undiff_vs_Diff.out$FDR <= 0.1] <- "Undifferentiated"
BE_Undiff_vs_Diff.out$specificity[BE_Undiff_vs_Diff.out$logFC > 0 & BE_Undiff_vs_Diff.out$FDR <= 0.1] <- "Foveolar_differentiated"

out$BE_Undiff_vs_Diff <- BE_Undiff_vs_Diff.out

# DE within undiff cells tissue
sce_test <- cur_sce[,cur_sce$cell_type == "Undifferentiated"]
Undiff_BE_vs_GC <- DE.edgeR(sce_test, conditions = sce_test$Tissue,
                              covariate = sce_test$Patient, lfc = 0.5,
                              FDR = 1)

Undiff_BE_vs_GC.out <- rbind(Undiff_BE_vs_GC$GC,
                               Undiff_BE_vs_GC$BE[
              seq(nrow(Undiff_BE_vs_GC$BE), 1),])

Undiff_BE_vs_GC.out$ID <- rownames(Undiff_BE_vs_GC.out)

Undiff_BE_vs_GC.out$specificity <- NA
Undiff_BE_vs_GC.out$specificity[Undiff_BE_vs_GC.out$logFC < 0 & Undiff_BE_vs_GC.out$FDR <= 0.1] <- "GC"
Undiff_BE_vs_GC.out$specificity[Undiff_BE_vs_GC.out$logFC > 0 & Undiff_BE_vs_GC.out$FDR <= 0.1] <- "BE"

out$Undiff_BE_vs_GC <- Undiff_BE_vs_GC.out

# DE within diff cells tissue
sce_test <- cur_sce[,cur_sce$cell_type == "Foveolar_differentiated"]
Diff_BE_vs_GC <- DE.edgeR(sce_test, conditions = sce_test$Tissue,
                              covariate = sce_test$Patient, lfc = 0.5,
                              FDR = 1)

Diff_BE_vs_GC.out <- rbind(Diff_BE_vs_GC$GC,
                               Diff_BE_vs_GC$BE[
              seq(nrow(Diff_BE_vs_GC$BE), 1),])

Diff_BE_vs_GC.out$ID <- rownames(Diff_BE_vs_GC.out)

Diff_BE_vs_GC.out$specificity <- NA
Diff_BE_vs_GC.out$specificity[Diff_BE_vs_GC.out$logFC < 0 & Diff_BE_vs_GC.out$FDR <= 0.1] <- "GC"
Diff_BE_vs_GC.out$specificity[Diff_BE_vs_GC.out$logFC > 0 & Diff_BE_vs_GC.out$FDR <= 0.1] <- "BE"

out$Diff_BE_vs_GC <- Diff_BE_vs_GC.out


# Visualize genes 
cur_tab <- out$Diff_BE_vs_GC
cur_mat <- logcounts(cur_sce)[c(rownames(cur_tab)[1:20], rownames(cur_tab)[(nrow(cur_tab)-20):nrow(cur_tab)]),]
colnames(cur_mat) <- paste(cur_sce$Barcode, cur_sce$Patient)
pheatmap(cur_mat, 
         cluster_rows = FALSE, 
         annotation_col = data.frame(row.names = colnames(cur_mat),
                                     cell_type = cur_sce$cell_type,
                                     tissue = cur_sce$Tissue),
         show_colnames = FALSE, gaps_row = 20)

# Save results
write.xlsx(out, file = "../../../Dropbox (Personal)/Origin_of_BE_draft/Figures/Supplemental_tables/Table_S7.xlsx")
```


```{r}
# Find overlap between Undiff and Diff specific genes and tissue comparisons
# We save the comparison between the tissues
# We then remove unecessary columns and add the comparison between undiff and diff

# List to save results
out_spec <- list()

# Undifferentiated
# GC specific
cur_1 <- out$GC_Undiff_vs_Diff[out$GC_Undiff_vs_Diff$specificity == "Undifferentiated" &
                                 !is.na(out$GC_Undiff_vs_Diff$specificity),]
cur_2 <- out$Undiff_BE_vs_GC[out$Undiff_BE_vs_GC$specificity == "GC" &
                               !is.na(out$Undiff_BE_vs_GC$specificity),]
shared.genes <- intersect(rownames(cur_1), rownames(cur_2))
GC_undiff <- cbind(out$GC_Undiff_vs_Diff[shared.genes,], out$Undiff_BE_vs_GC[shared.genes,])
GC_undiff <- GC_undiff[,c(1,5,9,13,14,15)]

colnames(GC_undiff)[1:4] <- c(paste(colnames(GC_undiff)[1:2], "Undiff_vs_Diff", sep = "."),
                  paste(colnames(GC_undiff)[3:4], "BE_vs_GC", sep = "."))

# Order table
GC_undiff <- GC_undiff[order(rowMeans(GC_undiff[,c(2,4)]), decreasing = FALSE),]
out_spec$GC_undiff_specific <- GC_undiff

# BE specific
cur_1 <- out$BE_Undiff_vs_Diff[out$BE_Undiff_vs_Diff$specificity == "Undifferentiated" &
                                 !is.na(out$BE_Undiff_vs_Diff$specificity),]
cur_2 <- out$Undiff_BE_vs_GC[out$Undiff_BE_vs_GC$specificity == "BE" &
                               !is.na(out$Undiff_BE_vs_GC$specificity),]
shared.genes <- intersect(rownames(cur_1), rownames(cur_2))
BE_undiff <- cbind(out$BE_Undiff_vs_Diff[shared.genes,], out$Undiff_BE_vs_GC[shared.genes,])
BE_undiff <- BE_undiff[,c(1,5,9,13,14,15)]

colnames(BE_undiff)[1:4] <- c(paste(colnames(BE_undiff)[1:2], "Undiff_vs_Diff", sep = "."),
                  paste(colnames(BE_undiff)[3:4], "BE_vs_GC", sep = "."))

# Order table
BE_undiff <- BE_undiff[order(rowMeans(BE_undiff[,c(2,4)]), decreasing = FALSE),]
out_spec$BE_undiff_specific <- BE_undiff

# Differentiated
# GC specific
cur_1 <- out$GC_Undiff_vs_Diff[out$GC_Undiff_vs_Diff$specificity == "Foveolar_differentiated" &
                                 !is.na(out$GC_Undiff_vs_Diff$specificity),]
cur_2 <- out$Diff_BE_vs_GC[out$Diff_BE_vs_GC$specificity == "GC" &
                               !is.na(out$Diff_BE_vs_GC$specificity),]
shared.genes <- intersect(rownames(cur_1), rownames(cur_2))
GC_diff <- cbind(out$GC_Undiff_vs_Diff[shared.genes,], out$Diff_BE_vs_GC[shared.genes,])
GC_diff <- GC_diff[,c(1,5,9,13,14,15)]

colnames(GC_diff)[1:4] <- c(paste(colnames(GC_diff)[1:2], "Undiff_vs_Diff", sep = "."),
                  paste(colnames(GC_diff)[3:4], "BE_vs_GC", sep = "."))

# Order table
GC_diff <- GC_diff[order(rowMeans(GC_diff[,c(2,4)]), decreasing = FALSE),]
out_spec$GC_diff_specific <- GC_diff

# BE specific
cur_1 <- out$BE_Undiff_vs_Diff[out$BE_Undiff_vs_Diff$specificity == "Foveolar_differentiated" &
                                 !is.na(out$BE_Undiff_vs_Diff$specificity),]
cur_2 <- out$Diff_BE_vs_GC[out$Diff_BE_vs_GC$specificity == "BE" &
                               !is.na(out$Diff_BE_vs_GC$specificity),]
shared.genes <- intersect(rownames(cur_1), rownames(cur_2))
BE_diff <- cbind(out$BE_Undiff_vs_Diff[shared.genes,], out$Diff_BE_vs_GC[shared.genes,])
BE_diff <- BE_diff[,c(1,5,9,13,14,15)]

colnames(BE_diff)[1:4] <- c(paste(colnames(BE_diff)[1:2], "Undiff_vs_Diff", sep = "."),
                  paste(colnames(BE_diff)[3:4], "BE_vs_GC", sep = "."))

# Order table
BE_diff <- BE_diff[order(rowMeans(BE_diff[,c(2,4)]), decreasing = FALSE),]
out_spec$BE_diff_specific <- BE_diff

# Visualize results
cur_mat <- logcounts(cur_sce)[out_spec$GC_diff_specific$ID,]
colnames(cur_mat) <- paste(cur_sce$Barcode, cur_sce$Patient)
cur_dat <- data.frame(row.names = colnames(cur_mat),
                                     cell_type = cur_sce$cell_type,
                                     tissue = cur_sce$Tissue)
pheatmap(cur_mat[,order(cur_dat$cell_type, cur_dat$tissue)], 
         cluster_rows = FALSE, 
         annotation_col = cur_dat[order(cur_dat$cell_type, cur_dat$tissue),],
         show_colnames = FALSE, cluster_cols = FALSE)

# Save results
write.xlsx(out_spec, file = "../../../Dropbox (Personal)/Origin_of_BE_draft/Figures/Supplemental_tables/Table_S8.xlsx")
```

# Visualize marker genes

```{r}
genes <- c("MYC", "HNF4A", "KRT7", "NCL", "CD44", "CLDN3", 
           "MGST2", "MGST3", "CHCHD10", "DNPH1", 
           "FBP1", "OCIAD2", "LEFTY1")

# Normalization across all cells displayed
cur_sce <- computeSumFactors(cur_sce, clusters=paste(cur_sce$Tissue, cur_sce$cell_type))
cur_sce <- normalize(cur_sce, return_log = TRUE)

# Visualize the size factors
ggplot(data.frame(UMAP1 = UMAP$layout[,1],
                  UMAP2 = UMAP$layout[,2],
                  sf = sizeFactors(cur_sce))) +
  geom_point(aes(UMAP1, UMAP2, colour = sf)) +
  scale_colour_viridis()

# Visualize library size
ggplot(data.frame(UMAP1 = UMAP$layout[,1],
                  UMAP2 = UMAP$layout[,2],
                  size = log10(colSums(counts(cur_sce))))) +
  geom_point(aes(UMAP1, UMAP2, colour = size)) +
  scale_colour_viridis()

# Visualize gene expression in form of boxplots
for.plot <- logcounts(cur_sce)[match(genes, rowData(cur_sce)$Symbol),]
colnames(for.plot) <- paste(cur_sce$Tissue, cur_sce$cell_type, cur_sce$Barcode, sep = "_")

# Order cells
ind_order <- c(which(cur_sce$Tissue == "GC" & cur_sce$cell_type == "Undifferentiated"),
               which(cur_sce$Tissue == "GC" & cur_sce$cell_type == "Foveolar_Intermediate"),
               which(cur_sce$Tissue == "GC" & cur_sce$cell_type == "Foveolar_differentiated"),
               which(cur_sce$Tissue == "BE" & cur_sce$cell_type == "Undifferentiated"),
               which(cur_sce$Tissue == "BE" & cur_sce$cell_type == "Dividing"),
               which(cur_sce$Tissue == "BE" & cur_sce$cell_type == "Foveolar_Intermediate"),
               which(cur_sce$Tissue == "BE" & cur_sce$cell_type == "Foveolar_differentiated"))

# Heatmap
dev.off()
pdf("../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/Heatmap_marker_genes.pdf", 
    width = 10, height = 5)
pheatmap(for.plot[,ind_order], cluster_rows = FALSE, cluster_cols = FALSE,
         color = viridis(100), labels_row = genes, 
         annotation_col = data.frame(row.names = colnames(for.plot)[ind_order],
                                     cell_type = cur_sce$cell_type[ind_order],
                                     tissue = cur_sce$Tissue[ind_order]),
         annotation_colors = list(cell_type = colour_vector,
                                  tissue = metadata(cur_sce)$colour_vector),
         show_colnames = FALSE)
dev.off()

pdf("../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/Heatmap_marker_genes.scaled.pdf", 
    width = 10, height = 5)
pheatmap(for.plot[,ind_order], cluster_rows = FALSE, cluster_cols = FALSE,
         color = colorRampPalette(c("dark blue", "blue", "white", "red", "dark red"))(100), 
         labels_row = genes, scale = "row",
         annotation_col = data.frame(row.names = colnames(for.plot)[ind_order],
                                     cell_type = cur_sce$cell_type[ind_order],
                                     tissue = cur_sce$Tissue[ind_order]),
         annotation_colors = list(cell_type = colour_vector,
                                  tissue = metadata(cur_sce)$colour_vector),
         show_colnames = FALSE)
dev.off()
```

# Calculate pseudotime for GC and BE

Here, we compute the pseudotime for BE and GC to highlight the opposint HNF4 and MYC signalling.

```{r}
# Select GC
cur_GC <- cur_sce[,cur_sce$Tissue == "GC"]
cur_GC <- cur_GC[Matrix::rowMeans(counts(cur_GC)) > 0,]

# Don't rescale to keep BE and GC comparable
#cur_GC <- normalize(cur_GC)

# Pseudotime 
pca_GC <- t(corrected[,cur_sce$Tissue == "GC"])
rownames(pca_GC) <- paste(cur_GC$Tissue, cur_GC$Patient, cur_GC$Barcode, sep = "_") 
colnames(pca_GC) <- paste("PC", seq(1,ncol(pca_GC)), sep = "")
clusters <- cur_GC$cell_type
names(clusters) <- rownames(pca_GC)
pt_GC <- PT(rd = pca_GC[,1:3], clusters = clusters, 
            col_vector = colour_vector[cur_GC$cell_type])

# Visualize trajectory
set.seed(12345)
rand <- rnorm(n = ncol(cur_GC), mean = 0, sd = 0.1)
GC.PT <- ggplot(data.frame(PT = pt_GC[,"lambda"],
                  value = rand,
                  cell_type = cur_GC$cell_type)) +
  geom_point(aes(PT, value, fill = cell_type), shape = 21, size = 2) +
  scale_fill_manual(values = colour_vector)

cur_df <- data.frame(PT = pt_GC[,"lambda"],
                  value = rand,
                  MYC = logcounts(cur_GC)[rowData(cur_GC)$Symbol == "MYC",])
GC_MYC <- ggplot() +
  geom_point(data = cur_df[cur_df$MYC == 0,], 
             aes(PT, value, colour = MYC), size = 2) +
    geom_point(data = cur_df[cur_df$MYC > 0,], 
             aes(PT, value, colour = MYC), size = 2) +
  scale_colour_viridis(limits = c(0, 4.2))

cur_df <- data.frame(PT = pt_GC[,"lambda"],
                  value = rand,
                  HNF4A = logcounts(cur_GC)[rowData(cur_GC)$Symbol == "HNF4A",])

GC_HNF4A <- ggplot() +
  geom_point(data = cur_df[cur_df$HNF4A == 0,], 
             aes(PT, value, colour = HNF4A), size = 2) +
    geom_point(data = cur_df[cur_df$HNF4A > 0,], 
             aes(PT, value, colour = HNF4A), size = 2) +
  scale_colour_viridis(limits = c(0, 4.2))

ggsave(filename = "../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/GC_pseudotime.pdf", 
       plot = GC.PT, width = 7, height = 2)
ggsave(filename = "../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/GC_MYC.pdf", 
       plot = GC_MYC, width = 7, height = 2)
ggsave(filename = "../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/GC_HNF4A.pdf", 
       plot = GC_HNF4A, width = 7, height = 2)

# Select BE
cur_BE <- cur_sce[,cur_sce$Tissue == "BE"]
cur_BE <- cur_BE[Matrix::rowMeans(counts(cur_BE)) > 0,]

# Don't rescale to keep BE and GC comparable
#cur_BE <- normalize(cur_BE)

# Pseudotime 
pca_BE <- t(corrected[,cur_sce$Tissue == "BE"])
rownames(pca_BE) <- paste(cur_BE$Tissue, cur_BE$Patient, cur_BE$Barcode, sep = "_") 
colnames(pca_BE) <- paste("PC", seq(1,ncol(pca_BE)), sep = "")
clusters <- cur_BE$cell_type
names(clusters) <- rownames(pca_BE)
pt_BE <- PT(rd = pca_BE[,1:3], clusters = clusters, 
            col_vector = colour_vector[cur_BE$cell_type])

# Visualize trajectory
set.seed(12345)
rand <- rnorm(n = ncol(cur_BE), mean = 0, sd = 0.1)
BE.PT <- ggplot(data.frame(PT = pt_BE[,"lambda"],
                  value = rand,
                  cell_type = cur_BE$cell_type)) +
  geom_point(aes(PT, value, fill = cell_type), shape = 21, size = 2) +
  scale_fill_manual(values = colour_vector)

cur_df <- data.frame(PT = pt_BE[,"lambda"],
                  value = rand,
                  MYC = logcounts(cur_BE)[rowData(cur_BE)$Symbol == "MYC",])
BE_MYC <- ggplot() +
  geom_point(data = cur_df[cur_df$MYC == 0,], 
             aes(PT, value, colour = MYC), size = 2) +
    geom_point(data = cur_df[cur_df$MYC > 0,], 
             aes(PT, value, colour = MYC), size = 2) +
  scale_colour_viridis(limits = c(0, 4.2))

cur_df <- data.frame(PT = pt_BE[,"lambda"],
                  value = rand,
                  HNF4A = logcounts(cur_BE)[rowData(cur_BE)$Symbol == "HNF4A",])
BE_HNF4A <- ggplot() +
  geom_point(data = cur_df[cur_df$HNF4A == 0,], 
             aes(PT, value, colour = HNF4A), size = 2) +
    geom_point(data = cur_df[cur_df$HNF4A > 0,], 
             aes(PT, value, colour = HNF4A), size = 2) +
  scale_colour_viridis(limits = c(0, 4.2))

ggsave(filename = "../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/BE_pseudotime.pdf", 
       plot = BE.PT, width = 7, height = 2)
ggsave(filename = "../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/BE_MYC.pdf", 
       plot = BE_MYC, width = 7, height = 2)
ggsave(filename = "../../../Dropbox (Personal)/Oesophagus_single_cell/Results/Figures/Figure_3/BE_HNF4A.pdf", 
       plot = BE_HNF4A, width = 7, height = 2)

```



